### PR TITLE
Client-side mods

### PR DESCRIPTION
**Pull request by duckbrain**
This adds client-side scripting to minetest. Scripts located in the `client` directory are transferred with media, and an `init.lua` file is run for each, mod. Client-side mods are always run in the secure environment and are only able to access the builtin scripts and the in-memory scripts transferred from the server.

In the current state, these mods are not useful (unless you only want to print to the user's terminal), but it is ready to have API's added that allow client-side prediction, input handling, messaging with the server, and many other features.

The benefit to getting this merged early, is it allows others to start working on API's.

Part of #3440 

~~This is WIP, because I realized changing the in-memory modpath, currently in the format of `<modname>/<filename.lua>` to `<modname>/client/<filename.lua>` would allow code like the following to work on the server or the client:~~

~~dofile(core.get_modpath("modtest") .. "/client/secondary-file.lua")~~

### CODE REVIEW

**Comment by sofar** [comment]
I'm fairly confused, I see a lot of stub methods for things that I don't think should ever be in a client script library: e.g. ban player methods. Are these copied over from the server side?

I'd expect the client API to be vastly different and almost entirely incompatible - e.g. `get_node(pos)` may just succeed, but `set_node(pos, node)` would surely be prohibited or nonexistant in a client API.

I can see server and client sharing some code bits (e.g. data structures/objects that don't modify the world), though.

**Comment by duckbrain** [comment]
That is copied from the server. The idea behind duplicating the server API is giving the client everything it needs to do client-side prediction. With the concept of placing blocks, the block would appear or not appear immediately based on what the server is going to tell it (if possible). The more API's that exist in common the more code that can be easily ported.

If this is not desired, that certainly makes things easier to do.

**Comment by nerzhul** [comment]
i don't agree with this design this is not secure for client.

First i think we should expose a client side API to customize GUI, next add a "store" with trusted and maybe verified mods. This API should be very restrained client side

**Comment by duckbrain** [comment]
@nerzhul I'm trying to understand the security concern. Is it that having a large API causes a larger attack surface or that having the server send code over the network is dangerous?

I have no problem with a smaller API; it just seems to me that different people (including me) have a different interpretation of what they want a client-side API to do, and some consensus would be nice.

As far as having a centralized "store" or repository of mods, I think that will cause more problems than it solves. The only advantage I see that it gives us is being able to moderate mods, but in such a system, creating a new account and uploading a malicious mod again under a different name would be pretty easy. I'm not under the impression we have the resources to vet every mod update.

If there is a store, do we have to tell the user that they need to have certain mods installed before connecting or do we download them while connecting? This also could make some mods not work (at least partially) without an internet connection, and creates issues with matching versions between client and server, especially while developing.

> The general design should be much like the Web browser-JavaScript relationship. During the connection, the server sends all code that should be executed to the client, which runs it inside a locked down sandbox. Unlike with server mods, there will be no way (without changing C++ source code) to execute code outside of that sandbox.
>
> [Minetest Wiki: Client scripting plans](http://dev.minetest.net/Client_scripting_plans#General_design)

@sofar The more I look into it, my dream of being able to move code into a `client/` directory, include it from the server side code, and have it work 90% of the time, seems impractical. Who would have thunk?

If you think it's appropriate, I'll just remove all the non-essential client-side API's (basically everything except `print`, `get_modpath`, and `dofile`), and we can use this as a base to implement what will actually be used.

**Comment by nerzhul** [comment]
@duckbrain in fact client side should have a dedicated lua stack for mods separated from the whole stack permitting players to import mods themselves to modify gui etc. It's the first goal. Permitting server to send mods is not a good start because of security. If i add rogue mod i can do things on player's computer

**Comment by rubenwardy** [comment]
> If i add rogue mod i can do things on player's computer

Sandboxing. See Html and javascript

> Permitting server to send mods is not a good start because of security.

Servers need to be able to send mods as the most useful types of client mods are prediction and gui, both of which are mod/subgame related

**Comment by sfan5** [comment]
IMO client-side modding should be made completely from scratch and not just mainly copied from the server API. Some methods may occurr in both APIs but the vast majority of lua_api.txt is only useful on the server.

**Comment by paramat** [comment]
You might want to chat to celeron55 about this as he has intentions for this feature.

**Comment by sofar** [comment]
I'm not necessarily against the server sending mods to the client, but I agree with @nerzhul that client-side mods should be able to get loaded by the end user.

Client-side prediction is one of the least interesting things client-side modding can do. I actually think client side mods should never attempt to modify (clients perception of) blocks to begin with, since the risk of unexpected behavior is very large, and any breakage will make things confusing for players.

Focus needs to be on:
- HUD, including chat
- sound effects, ambiance
- particles and visual effects
- reacting to (game) events (blocks placed, damage events) with non-game changing effects

**Comment by sofar** [comment]
I like to draw inspiration from mods written for World of Warcraft. These mods are entirely "ephemereal" in the sense that they don't influence the events happening in the world, but everyone who used them knows that they make a significant impact on the perception of the game. So much that the stock game basically uses them for large parts of the basic game functionality.

In wow the mods don't ship from the server. but mods can interact with other users' mods through communication channels (messages) and exchange data (e.g. damage counters, or resources discovered, or threat meters) and visually represent this in the game (e.g. modify the minimap, display a warning, play a sound). Incidentally, wow mods are entirely written in Lua.

I'd love to see the `ambiance` mod replaced with client-side modding entirely - it reacts to game changes (nodes placed, entities punching, velocities changing perhaps, even - a fall halted, stuff like that) and plays sounds in the right locations in the game, and perhaps add particle effects (smoke over torches, walking/digging particles). Stuff like that.

**Comment by nerzhul** [comment]
@sofar i agree with the World of warcraft modding design. I think this should be the correct behaviour -------- Message original --------Objet : Re: [minetest/minetest] WIP: Client-side mods (#4577)De : Auke Kok À : minetest/minetest Cc : nerzhul ,Mention I like to draw inspiration from mods written for World of Warcraft. These mods are entirely "ephemereal" in the sense that they don't influence the events happening in the world, but everyone who used them knows that they make a significant impact on the perception of the game. So much that the stock game basically uses them for large parts of the basic game functionality.

In wow the mods don't ship from the server. but mods can interact with other users' mods through communication channels (messages) and exchange data (e.g. damage counters, or resources discovered, or threat meters) and visually represent this in the game (e.g. modify the minimap, display a warning, play a sound). Incidentally, wow mods are entirely written in Lua.

I'd love to see the ambiance mod replaced with client-side modding entirely - it reacts to game changes (nodes placed, entities punching, velocities changing perhaps, even - a fall halted, stuff like that) and plays sounds in the right locations in the game, and perhaps add particle effects (smoke over torches, walking/digging particles). Stuff like that.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/minetest/minetest","title":"minetest/minetest","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/minetest/minetest"}},"updates":{"snippets":[{"icon":"PERSON","message":"@sofar in #4577: I like to draw inspiration from mods written for World of Warcraft. These mods are entirely \"ephemereal\" in the sense that they don't influence the events happening in the world, but everyone who used them knows that they make a significant impact on the perception of the game. So much that the stock game basically uses them for large parts of the basic game functionality.\r\n\r\nIn wow the mods don't ship from the server. but mods can interact with other users' mods through communication channels (messages) and exchange data (e.g. damage counters, or resources discovered, or threat meters) and visually represent this in the game (e.g. modify the minimap, display a warning, play a sound). Incidentally, wow mods are entirely written in Lua.\r\n\r\nI'd love to see the `ambiance` mod replaced with client-side modding entirely - it reacts to game changes (nodes placed, entities punching, velocities changing perhaps, even - a fall halted, stuff like that) and plays sounds in the right locations in the game, and perhaps add particle effects (smoke over torches, walking/digging particles). Stuff like that."}],"action":{"name":"View Pull Request","url":"https://github.com/minetest/minetest/pull/4577#issuecomment-252020323"}}}

**Comment by Fixer-007** [comment]
I remember in CS 1.X with metamod you can load totally different bots, and they will fight properly (you can make them per team or mix in any way), amazingly it works. At least from damage/position standpoint, they are not reacting to messages from other bot though. Not sure if it same as in WOW.

**Comment by sofar** [comment]
@Fixer-007 bots sounds exactly like NPC's or mobs (and those are server-side in minetest).

**Comment by duckbrain** [comment]
NPC's, mobs, vehicles, and powerups were the sorts of things I was thinking of with client-side scripting. Usings boats as an example, having the attachment and controls occur on the client-side could really help with sluggishness of networks/servers. It could make combat with mobs a little more enjoyable as well.

**Comment by nerzhul** [comment]
@DonBatman this is another time a wrong pattern, attachment should be only an attribute to attach a model on player, not a complete entity like now.

**Comment by red-001** [comment]
If this gets merged could it get its own branch?

**Comment by celeron55** [comment]
See my client_player_physics_script branch for one useful client-side API.

The script loading is not based on this PR though.

**Comment by est31** [comment]
Some of my thoughts on client side mods (mostly security related):

* matching the server API to 100% is not really useful
* Mod security should not be possible to be turned off -- (I think it seems this is already the case, just mentioning for completion)
* While reading file I/O should be okay for script files, writing I/O should never be possible. So the API for clientside scripting needs to be even more restricted than the secure server API
* For clientside scripting, not even builtin should get a non secure environment
* No lua voxel manips should be allowed, as they don't perform index checks (due to performance considerations), at least AFAIK, thus exposing an arbitrary memory write vulnerability which can probably be escalated to RCE 
* In the best case all clientside scripting stuff is put in a separate process, which can be sandboxed later on using the OS's sandboxing mechanisms (cgroups, etc). This doesn't mean that everything has to happen asynchronously, we can very well just simply block when lua performs some computation, and vice versa.

**Comment by nerzhul** [comment]
@est31 i agree with you about security but i don't think Linux sandboxing (cgroups) is possible because this is a pure userspace process and Lua should interact with irrlicht GUI thread

File: builtin/client/init.lua [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+local commonpath = scriptpath.."common"..DIR_DELIM
+local gamepath = scriptpath.."game"..DIR_DELIM
+
+dofile(commonpath.."vector.lua")
+
+dofile(gamepath.."constants.lua")
+dofile(gamepath.."item.lua")
+
+if core.setting_getbool("mod_profiling") then
+	dofile(gamepath.."mod_profiling.lua")
```
Comment (t4im) [inline builtin/client/init.lua]: That doesn't exist since 3 months anymore.

**Comment by ShadowNinja** [comment]
Client side Lua mods will not necessarily be stored on disk (although they may be for caching) therefore the regular `loadfile`/`dofile` API for loading secondary files won't work (I'm talking about, e.g., loading `client/extras.lua` from `client/init.lua`).  Instead you'll have to add a function like `load_script("extras.lua", ...)` which returns the loaded Lua chunk or the file as a string or file-like object.

There should be no access to anything outside of Minetest.  That means no filesystem access of any kind.  Otherwise mods may be able to read sensitive information (e.g. `~/.ssh/id_*`).  Scripts will have to have access to the Minetest environment and GUI, and they will need a method to communicate with the server scripts.

Something to consider is that without TLS/DTLS the code could be modified in transit.

**Comment by rubenwardy** [comment]
Dofile should be overriden to use a virtual filesystem

File: src/script/cpp_api/s_security.cpp [hunk 2]
```diff
@@ -257,6 +259,20 @@ bool ScriptApiSecurity::isSecure(lua_State *L)
...
+	lua_pop(L, 1);
+	
+	Client *client = script->getClient();
+	
+	if (client) {
+		ClientScripting *s = client->getScripting();
+		if (s->contains(path)) {
+			std::string fileContents = (*s)[path];
+			return !luaL_loadbuffer(L, fileContents.c_str(), fileContents.length(), NULL);
+		}
```
Comment (duckbrain) [inline src/script/cpp_api/s_security.cpp]: `ScriptApiSecurity::checkPath` checks if it is a client script and limits it to the builtin directory.

File: src/script/cpp_api/s_security.cpp [hunk 3]
```diff
@@ -257,6 +259,20 @@ bool ScriptApiSecurity::isSecure(lua_State *L)
 
 bool ScriptApiSecurity::safeLoadFile(lua_State *L, const char *path)
 {
+	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_SCRIPTAPI);
+	ScriptApiBase *script = (ScriptApiBase *) lua_touserdata(L, -1);
+	lua_pop(L, 1);
+	
+	Client *client = script->getClient();
```
Comment (duckbrain) [inline src/script/cpp_api/s_security.cpp]: I don't understand this request.

**Comment by duckbrain** [comment]
@ShadowNinja and @rubenwardy This is already implemented, the only exception is the builtin directory.

@ShadowNinja The lack of encryption concerns me as well. Although it is theoretically a secure environment to run in, I'd be more comfortable with encryption on scripts and passwords at the least.

@celeron55 That sounds great. When I have some time I'll have to look about getting that merged in.

@est31 

> - matching the server API to 100% is not really useful

Yeah, after thinking about it more, I agree.

> - While reading file I/O should be okay for script files, writing I/O should never be possible. So the API for clientside scripting needs to be even more restricted than the secure server API
> - For clientside scripting, not even builtin should get a non secure environment

builtin gets to be run using dofile, it has no other privileges. (unless I missed something)

> - No lua voxel manips should be allowed, as they don't perform index checks (due to performance considerations), at least AFAIK, thus exposing an arbitrary memory write vulnerability which can probably be escalated to RCE

This is a compelling reason to not allow voxel manips

@all

Would the builtin scripts be enough to clear the environment of potentially dangerous functions or will a further change to the security sandbox be needed to lock this down more?

I've been busy lately, so I don't know how much more time I can devote to this. I'll try, but feel free to pull from and use this code as a starting point if you don't hear from me.

**Comment by raymoo** [comment]
:+1: to @ShadowNinja 's suggestion that we ban globals in the client-side API conventions

**Comment by est31** [comment]
Also, maybe put everything behind an option, so that this PR can be merged even though the implementation is not perfect, and we may mess around with it without stability concerns, especially considering we're having a 0.4.15 release soon.

File: src/script/cpp_api/s_security.cpp (hunk 3 above)
Comment (duckbrain) [inline src/script/cpp_api/s_security.cpp]: Are you asking to have this moved into the `if` statement below? The condition for the statement is dependant on this result.

**Comment by duckbrain** [comment]
Except for the change requested by @nerzhul that I did not understand, the changes have been made. I added the option @est31 suggested, and I think we could merge this in as an experiment.

**Comment by raymoo** [comment]
@duckbrain What @nerzhul wanted you to do was move the variable declaration + initialization into the condition of the if statement. This makes it so the if statement only executes when the result is not null, and makes the variable in scope in the body of the statement. I didn't understand this request the first time @nerzhul asked me to do it either.

**Comment by est31** [comment]
@duckbrain can you document the setting in settingtypes.txt and point out that it is unstable so may be changed or removed in future versions, or something like that?

**Comment by rubenwardy** [comment]
Please don't do that, it's harder to read and doesn't really have any benefit

**Comment by duckbrain** [comment]
@raymoo Thank you for clarifying.

@est31 Done and re-squashed. I hope
 
> Enable Client-side Mods \*Expiremental\* may be removed or changed in future versions

is okay for the friendly name. It's kind of long, but the "Experimental" is towards the beginning of the message.

@rubenwardy There is some arguable benefit of limiting the scope to where it is needed, but I'm with you on keeping it easier to read.

File: src/client.cpp [hunk 4]
```diff
@@ -1928,3 +1944,11 @@ scene::IAnimatedMesh* Client::getMesh(const std::string &filename)
 	smgr->getMeshCache()->removeMesh(mesh);
 	return mesh;
 }
+
+void Client::loadMods()
+{
+	if (!g_settings->getBool("enable_client_mods")) return;
```
Comment (nerzhul) [inline src/client.cpp]: two lines please

File: src/script/cpp_api/s_client.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+#ifndef S_CLIENT_H_
+#define S_CLIENT_H_
+
+#include "cpp_api/s_base.h"
+#include <set>
+
+class ScriptApiClient
+		: virtual public ScriptApiBase
```
Comment (nerzhul) [inline src/script/cpp_api/s_client.h]: please set it with upper line

File: src/client.cpp (hunk 4 above)
Comment (duckbrain) [inline src/client.cpp]: Done

File: src/script/cpp_api/s_client.h (hunk 5 above)
Comment (duckbrain) [inline src/script/cpp_api/s_client.h]: and done.

**Comment by PureTryOut** [comment]
>@est31: Also, maybe put everything behind an option, so that this PR can be merged even though the implementation is not perfect, and we may mess around with it without stability concerns, especially considering we're having a 0.4.15 release soon.

:+1: If this would allow mods to listen to keybindings as well, then please at least put it in behind an opt-in option. I would very much like to experiment with this.

**Comment by duckbrain** [comment]
@est31  It's already done. Just waiting on a merge.

File: builtin/client/readonly.lua [hunk 6]
```diff
@@ -0,0 +1,7 @@
+-- Set the global environment to a read-only mode.
+
+local f=function (t,i) error("cannot redefine global variable `"..i.."'",2) end
```
Comment (SmallJoker) [inline builtin/client/readonly.lua]: You could put this function directly into `setmetatable`.
Also the code style is not consistent with the other builtin code.

File: builtin/settingtypes.txt [hunk 7]
```diff
@@ -1274,3 +1274,6 @@ modstore_details_url (Modstore details URL) string https://forum.minetest.net/mm
 
 #    Print the engine's profiling data in regular intervals (in seconds). 0 = disable. Useful for developers.
 profiler_print_interval (Engine profiling data print interval) int 0
+
+# Enable the expiremental client-side mods API
+enable_client_mods (Enable Client-side Mods *Expiremental* may be removed or changed in future versions) bool false
```
Comment (SmallJoker) [inline builtin/settingtypes.txt]: `experimental` spelled wrong two times

File: src/client.cpp [hunk 8]
```diff
@@ -743,6 +745,20 @@ bool Client::loadMedia(const std::string &data, const std::string &filename)
 		return true;
 	}
 
+	const char *script_ext[] = {
+		".lua",
+		NULL
+	};
+	name = removeStringEnd(filename, script_ext);
+	if(name != "")
```
Comment (SmallJoker) [inline src/client.cpp]: Bad code style shouldn't be adapted. [Guidelines](http://dev.minetest.net/Code_style_guidelines)

**Comment by duckbrain** [comment]
I think the failed built has something to do with an existing issue unrelated to this pull request. See: #3786 and #3940

The changes requested by @SmallJoker have been implemented, except the `readonly.lua` file was removed as it was not included in any other scripts and I could not get it to work.

@nerzhul's undone change seemed to be unwanted by @rubenwardy and myself.

**Comment by paramat** [comment]
Would you like me to remove the WIP label?

**Comment by duckbrain** [comment]
Yes please. Thanks.

File: src/script/cpp_api/s_security.cpp (hunk 3 above)
Comment (Foghrye4) [inline src/script/cpp_api/s_security.cpp]: @duckbrain I believe nerzhul means:

```
if (Client *client = script->getClient()) {
```

**Comment by paramat** [comment]
@celeron55 and @kwolekr ('hmmmm' on IRC) were planning at one point to add clientside scripting, it would be a good idea to get their reviews on this.

**Comment by est31** [comment]
I doubt that this is the final design for client side scripting, but at least having it would be a start...

**Comment by rubenwardy** [comment]
I don't like that this just shares things from builtin/game/. If you are to do this, you should add builtin/server/ then use **everything** from builtin/game/, and only /server/ on the server and /client/ on the client.

**Comment by duckbrain** [comment]
Rebased

**Comment by BluebirdGreycoat** [comment]
I would very much like to see this merged and behind an opt-in option.

As a server owner, I am greatly interested in the ability to upload code to the client (properly sandboxed, duh). MANY non-critical features of a game could benefit from being run on the client, thus freeing up the server for other more critical work.

I'm dreaming of being able to call `minetest.find_nodes_in_area()` on the client and use the information to play ambiance, including fire sounds and any other environment-related sound. We could scan for torches and add particles on the client without having to send particle spawner information over the network. And that's just the start ...

**Comment by paramat** [comment]
Should not be merged before release.

**Comment by duckbrain** [comment]
@BluebirdGreycoat This won't let you do that yet. It has the base of client-side modding, but practically no API. It would be of little use unless you want to print something.

I think it makes perfect sense to wait until after the release to merge this. I am, however, waiting to have a commit that won't need to be re-written to branch off of to start adding celeron55's calls to the API along with any other fun features.

When is release? I have 2 machines I work on this with, and rewriting Git history is a little bit of a pain to do repeatedly. I will squash and rebase (if needed) after release if it will be merged.

**Comment by paramat** [comment]
Release hopefully 21st Dec.

**Comment by ShadowNinja** [comment]
Some additional thoughts: we should consider running each mod in its own `lua_State` for sandboxing (the server uses some hacks to try to identify separate mods in a single state, but they don't completely work).

Also, security should be reviewed, and you might even want a completely separate more restrictive implementation for the client.  This absolutely **MUST** be secure.  In fact, given the consequences of making a mistake, I think it's important to look into defense in depth.  Something like running all Lua code in a separate process with `seccomp` could work, although it would have to be cross-platform.

Given the size of this project, maybe we should merge it into a separate branch, work on it there, and then merge it when it's ready.  We'd just have to make sure to actually finish it. :-)

**Comment by est31** [comment]
> Some additional thoughts: we should consider running each mod in its own `lua_State` for sandboxing

I don't know whether this will be advantageous... The big problem with sandboxing on the server lua is that you can have trusted and non trusted mods in your sandbox, but I think it doesn't make much sense to trust some mods coming from the server and not trust others.

> This absolutely MUST be secure. In fact, given the consequences of making a mistake, I think it's important to look into defense in depth. Something like running all Lua code in a separate process with seccomp could work, although it would have to be cross-platform.

I've suggested something similar above in the thread. I think its a good idea :) Still, we shouldn't require the author of this PR to do all the work for it.

We could start with a single platform implementation (like for GNU/Linux), and then expand to other OSes. Some OSes have very hard to do sandboxing, I've heard windows xp for example has. Either way, XP can be easily hacked with other means as well.

> Given the size of this project, maybe we should merge it into a separate branch, work on it there, and then merge it when it's ready. We'd just have to make sure to actually finish it. :-)

Alternatively, we could merge it but disabled by default. This also makes it easier for users to give feedback, and its the current approach taken by this PR. But its a gigantic project, I agree.

**Comment by duckbrain** [comment]
This has been squashed and rebased.

@ShadowNinja and @est31, I don't see much benefit of running each mod in a separate sandbox, but I agree with the idea of getting them to run in a separate process. The option to enable the client-side mods is there and off by default.

Although, security is of the utmost importance, this pull request is about creating a base on which to start building the API. Perhaps it would be better to discuss on the original issue #3440, or a new one.

**Comment by sofar** [comment]
Just to reiterate, the client-side API functions that have been requested (indirectly) most are:

- playing sounds (for ambiance, node sounds, fire sounds, etc.)
- creating particles (or particle spawners) (torch smoke, fire smoke etc.)
- updating and adding various HUD elements

These all are non-game changing functions, so they should be high priority and IMHO a top priority.

I do not think we should ever add entity manipulation support in client side modding, since that is just plain nonsense (player A's client moves mob B to pos C - packet to server to update pos -> packets from server to all other clients to update pos -> bullshit).

At most, one would allow client-side lua to do things like "on death of a mob make it play an animation instead of immediately removing the mesh" (non-game changing visual). However, that is a very marginal improvement over what can already be done right now (and just avoids the current broken design where mob entities are removed automatically when hp=0).

**Comment by est31** [comment]
Things like carts are quite laggy atm, and doing them on the client will remove that lag. Of course, it will create more problems with cheaters though.

**Comment by orwell96** [comment]
I would like to see object manipulation functions, BUT:
nothing is sent from client to server
Arriving server packets have priority

Use case: only predict what server will send anyway.
Example: trains/carts can be predicted to take a turn before the corresponding server packets arrive.
Or boats can be turned on player control pressed.
Main objective: prevent lag by running code synchronously on server and client.

**Comment by duckbrain** [comment]
My understanding is that player physics are computed on the client-side, and that entities can be attached to the player. Could this be an appropriate work-around for manipulating entities to create vehicles?

As long as there is [some sort of messaging from the client to server](http://dev.minetest.net/Client_scripting_plans) to indicate that a mod has loaded, mods could detect and still fall back to using the server-side calculated physics if the client does not support them. Otherwise, attach the entity to the player and [allow client-side mods to override player physics](https://github.com/minetest/minetest/pull/4577#issuecomment-252848192).

This does not solve the issues with carts with no one in them or mobs, but it is a start that doesn't require the "nonsense" @sofar mentions. However, I disagree that it is nonsense and still hold that it should be implemented.

@sofar Can you please explain why you are against this kind of API? The disadvantages I can come up with right now are a more chatty/complicated client/server protocol and more RAM/CPU usage on the server side. Are there other reasons against this?

I don't remember where I found this link, but it was somewhere associated with either a Minetest issue or the wiki: http://gabrielgambetta.com/fast_paced_multiplayer.html. I think it explains concepts @orwell96 referred to and how to deal with the threat of cheaters that @est31 mentioned. I thought it was a good read. This is why I originally wanted client-side mods to be somewhat compatible with server-side mods, but I realize that Minetest is not there yet.

**Comment by sofar** [comment]
I think it's nonsense since it won't actually improve the experience at all. A small amount of lag (0-200ms) is easily accepted by users, and from my experience the lag problem is much more that there is somehow way more lag in minetest servers than there should be(500ms-2000ms), and making an incredibly smart and complex system to work around this lag is the stupidest way to solve the actual problem of lag being too high. If you really want to invest years of fixing bugs in complex code, fix the reason why currently servers are drowning in seconds of lag for no apparent reason instead. For 200ms of lag, I can game world-wide. 2000ms lag should not exist anymore.

Another problem I have with that type of functionality is that it's going to be wholly unusable by mob mod makers - I haven't seen any mob API's out there that are IMHO capable of doing good movement prediction, since they make significant mistakes in the design (such as constantly re-updating the path calculation for no reason), and I have doubts that movement prediction is going to ever get used entirely.

Ultimately, wanting movement prediction is wanting your scurvy to hurt less, instead of eating fruits and vegetables, it's just avoiding the real problem.

Moreover, I have seen in many occasions that in `minetest_game` we've had to tone down significant visuals and audio effects because of the amount of network packets involved. Just look at how much the fire code has been redone. This is all stuff that makes a ton of sense to move to the client, since essentially it's the same thing as footstep sounds (which are already client side, and with client side lua could be moved to client-side lua as well!).

So if you want to make an API, and want to make it useful, quick, for things we have (1) a need for and (2) are incredibly useful and (3) provide an immediate improvement to the minetest experience, I know what I'd be doing first.

If the client-side lua is going to offer zero functionality until entity movement prediction is implemented, we're going to be waiting a long time.

so, it's a priority thing for me. I'd like to see footstep particles and node sounds and torch smoke, asap. The rest is more of an academic thing if you ask me.

**Comment by duckbrain** [comment]
Thank you, that makes sense.

I have no objection to moving sounds, particle effects, and HUD to the client side. That seems completely logical to me. I can see how that would be an improvement to network and server performance. _I_ have little interest for sounds, because I usually have my volume turned off, and I don't pay much attention to the particle effects, so I don't notice these as problems. The HUD improvements would be great addition. I agree that these should be made a priority.

Is there any objection to player physics being overridable? 100-200ms isn't really an acceptable amount of time for the player to respond to key presses, and this could offload more calculations from the server. I feel this could be used in a lot of creative and interesting ways to improve the game without adding any extra to the network chatter. I can make that a second pull request.

Also, is there anything I need to do to get this ready for merge? I want to make sure I'm not being waited on while I've got the time to do it.

**Comment by sofar** [comment]
> offload more calculations from the server 

I think this is a wrong assumption. You're likely actually going to do more calculations on the client, but the server still needs to do what it needs to do - including collision detection - otherwise it's just one exploitable mess.

**Comment by sofar** [comment]
> Also, is there anything I need to do to get this ready for merge? I want to make sure I'm not being waited on while I've got the time to do it.

I think this is good as-is if you'd ask me. It would be nice to have a concept mod that can be used to validate the functionality, as well as unit testing.

There may be a significant opportunity to move e.g. player footsteps to a client-side lua mod, not only to prove the concept but also to provide some significant template mod code that people can use to create new works from.

Plus, there's documentation. You may have to rename `lua_api.txt` to `server_lua_api.txt` and make a new `client_lua_api.txt` :)

**Comment by Fixer-007** [comment]
>I have no objection to moving sounds, particle effects, and HUD to the client side.

That looks like a good starting point.

>I have little interest for sounds, because I usually have my volume turned off

As a player I have big interest in sounds! There are too few of them right now and they slow down the damn server.

**Comment by sapier** [comment]
As soon as scripts are sent from server client side scripting **NEVER EVER** should have access to anything except minetest s own data. 
Thus you cannot use a full lua environemt as it's used on server, first step to do this is disable everything within lua not really required, e.g. lib loading, os access ... 

Starting point would be **not** loading luas default libs as it's done in ScriptApiBase::ScriptApiBase()  around Line 83

**Comment by rubenwardy** [comment]
Security should be added before this is merged - it is very important
It also stops breaking mods that start relying on insecure features

**Comment by est31** [comment]
I'm for requiring security before it gets enabled by default.

**Comment by sapier** [comment]
Looking at the recent commits where almost any security feature added has been disabled by default by some time I disagree with est31 (assuming his comment was meant as "merge it without security but disable it to compensate")

**Comment by nerzhul** [comment]
I let this PR opened some time to look at interesting things we can backport to #5088 but #5088 is now the official PR and minetest branch for this

**Comment by nerzhul** [comment]
supersede by #5088

**Pull request closed without merging**